### PR TITLE
MBIM QDU support for firmware downloading in the MM plugin

### PR DESCRIPTION
**Pull request by Jarvis-Jiang-G**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [x] Feature
- [ ] Documentation

This MR adds the mbim-qdu support for firmware downloading, it will be used for T99W175 module of Foxconn.

### CODE REVIEW

File: plugins/modem-manager/fu-mm-device.c [hunk 1]
```diff
@@ -653,6 +728,86 @@ fu_mm_device_write_firmware_qmi_pdc (FuDevice *device, GBytes *fw, GArray **acti
...
+	xb_builder_import_source (builder, source);
+	silo = xb_builder_compile (builder, XB_BUILDER_COMPILE_FLAG_NONE, NULL, error);
+	if (silo == NULL)
+		return FALSE;
+
+	/* get all the operation parts */
+	parts = xb_silo_query (silo, "parts/part", 0, error);
+	if (parts == NULL)
+		return FALSE;
+	part = g_ptr_array_index (parts, 0);
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:791]: I think you can use `xb_silo_query_first() `here

File: plugins/modem-manager/fu-mm-device.c [hunk 2]
```diff
@@ -653,6 +728,86 @@ fu_mm_device_write_firmware_qmi_pdc (FuDevice *device, GBytes *fw, GArray **acti
...
+	filename = xb_node_get_attr (part, "filename");
+	csum = xb_node_get_attr (part, "MD5");
+	data = fu_archive_lookup_by_fn (archive, filename, error);
+	if (data == NULL)
+		return FALSE;
+	csum_actual = g_compute_checksum_for_bytes (G_CHECKSUM_MD5, data);
+	if (g_strcmp0 (csum, csum_actual) != 0) {
+		g_debug ("MD5 not matched");
+		return FALSE;
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:800]: you need a `g_set_error` here if you're returning FALSE.

File: plugins/modem-manager/fu-mm-device.c [hunk 3]
```diff
@@ -679,6 +834,11 @@ fu_mm_device_write_firmware (FuDevice *device,
 	if (self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_QMI_PDC)
 		return fu_mm_device_write_firmware_qmi_pdc (device, fw, &self->qmi_pdc_active_id, error);
 
+	/* mbim qdu write operation */
+	if (self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_MBIM_QDU) {
+		return fu_mm_device_write_firmware_mbim_qdu (device, fw, error);
+	}
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:840]: no need for `{}` braces here

File: plugins/modem-manager/fu-mm-device.c [hunk 4]
```diff
@@ -752,7 +912,9 @@ fu_mm_device_attach (FuDevice *device, GError **error)
 
 	/* wait for re-probing after uninhibiting */
 	fu_device_set_remove_delay (device, FU_MM_DEVICE_REMOVE_DELAY_REPROBE);
-	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	if (!(self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_MBIM_QDU))
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:915]: I think `(self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_MBIM_QDU) == 0` would be clearer here

File: plugins/modem-manager/fu-plugin-modem-manager.c [hunk 5]
```diff
@@ -239,11 +256,17 @@ fu_plugin_mm_device_removed_cb (MMManager *manager, MMObject *modem, FuPlugin *p
...
+	MMModemFirmwareUpdateMethod dev_update_methods = fu_mm_device_get_update_methods(dev);
 	if (dev == NULL)
 		return;
 	g_debug ("removed modem: %s", mm_object_get_path (modem));
-	fu_plugin_cache_remove (plugin, object_path);
-	fu_plugin_device_remove (plugin, FU_DEVICE (dev));
+	mbim_mm_object_path = g_strdup (object_path);
+	
+	/* later removed the inhibited mbim device from the plugin */
+	if (!(dev_update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_MBIM_QDU)) {
```
Comment (hughsie) [inline plugins/modem-manager/fu-plugin-modem-manager.c:266]: `== 0` also please

File: plugins/modem-manager/fu-mm-utils.c [hunk 6]
```diff
@@ -46,6 +46,23 @@ fu_mm_utils_get_udev_port_info (GUdevDevice	 *device,
 		g_set_object (&parent, next);
 	}
 
+	if (!device_sysfs_path && g_strcmp0 (g_udev_device_get_subsystem (device), "wwan") == 0) {
+		parent = g_udev_device_get_parent (device);
+		while (parent != NULL) {
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-utils.c:51]: does `g_udev_device_get_parent_with_subsystem()` work better here?

File: plugins/modem-manager/fu-mm-device.c (hunk 1 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mm-device.c:791]: Yes, you are right, xb_silo_query_first() is more suitable here.

File: plugins/modem-manager/fu-mm-device.c (hunk 2 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mm-device.c:800]: Ok, I will add it, thanks.

File: plugins/modem-manager/fu-mm-device.c (hunk 4 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mm-device.c:915]: Ok, I will change it, thanks.

File: plugins/modem-manager/fu-mm-utils.c (hunk 6 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mm-utils.c:51]: Oh, I don't know, can you give me some suggestions?

File: plugins/modem-manager/fu-mbim-qdu-updater.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+#define __FU_MBIM_QDU_UPDATER_H
+
+#include <libmbim-glib.h>
+#include "fu-mm-device.h"
+
+#define FU_TYPE_MBIM_QDU_UPDATER (fu_mbim_qdu_updater_get_type ())
+G_DECLARE_FINAL_TYPE (FuMbimQduUpdater, fu_mbim_qdu_updater, FU, MBIM_QDU_UPDATER, GObject)
+
+FuMbimQduUpdater	*fu_mbim_qdu_updater_new 	(const gchar		*mbim_port);
+gboolean	 fu_mbim_qdu_updater_open	(FuMbimQduUpdater	*self,
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.h:17]: Alignment here is quite off, isn't it?

File: plugins/modem-manager/fu-mm-device.c [hunk 8]
```diff
@@ -116,6 +123,7 @@ validate_firmware_update_method (MMModemFirmwareUpdateMethod methods, GError **e
 	static const MMModemFirmwareUpdateMethod supported_combinations[] = {
 		MM_MODEM_FIRMWARE_UPDATE_METHOD_FASTBOOT,
 		MM_MODEM_FIRMWARE_UPDATE_METHOD_QMI_PDC | MM_MODEM_FIRMWARE_UPDATE_METHOD_FASTBOOT,
+		MM_MODEM_FIRMWARE_UPDATE_METHOD_MBIM_QDU,
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mm-device.c:126]: This update method does not exist in ModemManager. This pull request should have been sent for review after the corresponding update method is defined and used in ModemManager, otherwise, is this pull request only usable with some unknown custom version of MM?

Could you please push the MM changes first, we get them merged in MM, we then bump the MM dev version including the changes (e.g. to 1.17.1 or whatever) and then we build-depend on that version for this integration in fwupd?

Comment (hughsie) [reply plugins/modem-manager/fu-mm-device.c:126]: Yes, we definitely need the MM changes merged first -- good catch.

File: plugins/modem-manager/fu-mm-device.c [hunk 9]
```diff
@@ -275,6 +315,18 @@ fu_mm_device_probe_default (FuDevice *device, GError **error)
 		return FALSE;
 	}
 
+	/* Flag for mhi-based mbim device */
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mm-device.c:318]: I don't think the logic behind this flag is ok. Do you only need the is_mhi_dev flag to build a vendor ID with PCI prefix instead of USB prefix?

I think that instead of this logic we could use this patch I wrote last December, which we didn't integrate yet because we were waiting for the actual final setup to be ready, but it would probably be worth getting merged to make it work also with this new QDU method: https://github.com/aleksander0m/fwupd/commit/abf7685e8e233be01da4f677508f88a690ff7ecf

Could you see if that patch helps you to do the same?

Comment (aleksander0m) [reply plugins/modem-manager/fu-mm-device.c:318]: Also, this one may be worth including: https://github.com/aleksander0m/fwupd/commit/5e936040f21bc3084ddb4402d14457c1ff139b4c

File: plugins/modem-manager/fu-mm-device.c [hunk 10]
```diff
@@ -861,6 +1024,23 @@ fu_mm_device_udev_new (MMManager *manager,
 	fu_device_set_name (FU_DEVICE (self), info->name);
 	fu_device_set_version (FU_DEVICE (self), info->version);
 	self->update_methods = info->update_methods;
+	if (self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_MBIM_QDU) {
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mm-device.c:1027]: Is all this new block really needed? Why? At this point we're already in the upgrade process for the device and we can create the udev device with the info coming from the old MM device. If we needed to have vendor ids and protocol added, we could have added those items to the FuPluginMmInhibitedDeviceInfo but I don't really think you need them.

File: plugins/modem-manager/fu-mm-device.c [hunk 11]
```diff
@@ -897,6 +1077,13 @@ fu_mm_device_udev_add_port (FuMmDevice	*self,
 		return;
 	}
 
+	if (g_str_equal (subsystem, "wwan") &&
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mm-device.c:1080]: You're limiting the MBIM QDU download protocol exclusively to PCI devices now; is that expected? Is it possible to use this protocol through USB as well? Maybe we should also handle QMI ports in the wwan subsystem and MBIM ports in the usbmisc subsystem?

File: plugins/modem-manager/fu-mm-utils.c [hunk 12]
```diff
@@ -46,6 +46,23 @@ fu_mm_utils_get_udev_port_info (GUdevDevice	 *device,
 		g_set_object (&parent, next);
 	}
 
+	if (!device_sysfs_path && g_strcmp0 (g_udev_device_get_subsystem (device), "wwan") == 0) {
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mm-utils.c:49]: I added PCI support following the same logic as done in MM in this commit that I also commented about before: https://github.com/aleksander0m/fwupd/commit/abf7685e8e233be01da4f677508f88a690ff7ecf Could you see if that logic could fit better here? In that logic we look for the subsystem first with find_device_bus_subsystem() and then we apply different approaches to USB and PCI

File: plugins/modem-manager/fu-plugin-modem-manager.c [hunk 13]
```diff
@@ -162,10 +164,24 @@ fu_plugin_mm_udev_uevent_cb (GUdevClient	*udev,
 	if (g_strcmp0 (device_sysfs_path, priv->inhibited->physical_id) != 0)
 		return TRUE;
 
+	/* Remove the inhibited mbim device from the plugin once the first wwan device remove is detected */
+	if (g_str_equal (subsystem, "wwan") && g_str_equal (action, "remove") && mbim_mm_object_path != NULL) {
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-plugin-modem-manager.c:168]: Why is this needed?

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 14]
```diff
@@ -0,0 +1,545 @@
...
+	}
+ 
+	fu_mm_utils_get_port_info (self->mbim_port, &device_sysfs_path, NULL, NULL);
+	/* device will auto reboot after update */
+	/* pcie rescan is required */
+	/* mbim-proxy stuck due to mbim port suddenly lost */
+	if (device_sysfs_path) {
+		g_auto(GStrv) split = NULL;
+		split = g_strsplit (device_sysfs_path, "/", -1);
+		/* /sys/devices/pci0000:00/0000:00:1d.3/0000:02:00.0 */
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:194]: Isn't this very very specific to your own hardware?

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 15]
```diff
@@ -0,0 +1,545 @@
...
+			}
+
+			if (f)
+				fclose(f);
+
+			argc = g_new0 (gchar *, 4);
+			argc[3] = NULL;
+			argc[0] = g_strdup ("killall");
+			argc[1] = g_strdup ("-9");
+			argc[2] = g_strdup ("mbim-proxy");
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:222]: oh my god no

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 16]
```diff
@@ -0,0 +1,545 @@
...
+	WriteContext *ctx = user_data;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_write_response_parse (
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:303]: response can probably go to the previous line

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 17]
```diff
@@ -0,0 +1,545 @@
...
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_mbim_qdu_updater_file_write_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	MbimMessage *request  = NULL;
+	MbimMessage *response = NULL;
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:292]: g_autoptr() can be used for both response and request as well

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 18]
```diff
@@ -0,0 +1,545 @@
...
+		g_warn_if_fail (ctx->timeout_id == 0);
+		/* device will auto reboot right after update finish */
+		ctx->timeout_id = g_timeout_add_seconds (30, fu_mbim_qdu_updater_reboot_timeout, ctx);
+	}
+}
+
+static void
+fu_mbim_qdu_updater_file_open_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	MbimMessage *request = NULL;
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:339]: g_autoptr() for the MbimMessages

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 19]
```diff
@@ -0,0 +1,545 @@
...
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_open_response_parse (
+							response,
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:356]: response to the previous line

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 20]
```diff
@@ -0,0 +1,545 @@
...
+		g_debug ("error: couldn't parse response message: %s\n", error->message);
+		g_debug ("out_max_transfer_size: %d\n", (int)out_max_transfer_size);
+		g_debug ("out_max_window_size: %d\n", (int)out_max_window_size);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:369]: single whiteline to separate blocks

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 21]
```diff
@@ -0,0 +1,545 @@
...
+
+	total_sending_numbers = (g_bytes_get_size (ctx->blob)) / (out_max_transfer_size);
+	if ((g_bytes_get_size (ctx->blob)) % (out_max_transfer_size) != 0)
+		total_sending_numbers ++;
+
+	ctx->chunks = fu_chunk_array_new_from_bytes (ctx->blob,
+						0x00,	/* start addr */
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:376]: alignment seems off, isn't it?

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 22]
```diff
@@ -0,0 +1,545 @@
...
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_printerr ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_update_session_response_parse (
+							     response,
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:412]: response in previous line?

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 23]
```diff
@@ -0,0 +1,545 @@
...
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_write_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static void
+fu_mbim_qdu_updater_session_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	MbimMessage *response = NULL;
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:393]: g_autoptr

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 24]
```diff
@@ -0,0 +1,545 @@
...
+	MbimQduSessionType     out_current_session_type;
+	MbimQduSessionStatus   out_current_session_status;
+	MbimQduSessionType     out_last_session_type;
+	MbimQduSessionResult   out_last_session_result;
+	guint32                out_last_session_error_offset;
+	guint32                out_last_session_error_size;
+	g_autoptr(GError)      error = NULL;
+	MbimMessage *request = NULL;
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:401]: g_autoptr

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 25]
```diff
@@ -0,0 +1,545 @@
...
+	WriteContext *ctx = user_data;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_printerr ("error: operation failed: %s\n", error->message);
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:406]: I don't think there should be a single g_printerr() in all this patch, and there are multiple

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 26]
```diff
@@ -0,0 +1,545 @@
...
+							     response,
+							     &out_current_session_type,
+							     &out_current_session_status,
+							     &out_last_session_type,
+							     &out_last_session_result,
+							     &out_last_session_error_offset,
+							     &out_last_session_error_size,
+							     &error)) {
+		g_printerr ("error: couldn't parse response message: %s\n", error->message);
+		g_printerr ("out_current_session_type: %d\n", (int)out_current_session_type);
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:421]: You should only attempt to use the contents of the out parameters if mbim_message_qdu_update_session_response_parse succeeds. If the method fails, you can only read the error, and you must ignore all the other fields.

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 27]
```diff
@@ -0,0 +1,545 @@
...
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_open_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static void
+fu_mbim_qdu_updater_set_update_session (MbimDevice *device, WriteContext *ctx)
+{
+  	MbimMessage *request = NULL;
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:449]: g_autoptr

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 28]
```diff
@@ -0,0 +1,545 @@
...
+
+	request = (mbim_message_qdu_update_session_set_new (0, 1, &ctx->error));
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:451]: You can assume the set_new() always succeeds. Just pass a NULL for the GError and don't check for errors.

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 29]
```diff
@@ -0,0 +1,545 @@
...
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	g_debug ("[%s] Successfully request modem to update session", mbim_device_get_path_display (device));
+
+	request = (mbim_message_qdu_file_open_set_new (0, g_bytes_get_size (ctx->blob), &error));
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:436]: Why the extra () around the method call?
Also just pass NULL at the GError, the set_new() method should never fail.

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 30]
```diff
@@ -0,0 +1,545 @@
...
+	total_sending_numbers = (g_bytes_get_size (ctx->blob)) / (out_max_transfer_size);
+	if ((g_bytes_get_size (ctx->blob)) % (out_max_transfer_size) != 0)
+		total_sending_numbers ++;
+
+	ctx->chunks = fu_chunk_array_new_from_bytes (ctx->blob,
+						0x00,	/* start addr */
+						0x00,	/* page_sz */
+						out_max_transfer_size);
+	chk = g_ptr_array_index (ctx->chunks, 0);
+	request = (mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), &error));
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:380]: Just pass NULL at the GError, the set_new() method will never fail

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 31]
```diff
@@ -0,0 +1,545 @@
...
+		return;
+	}
+
+	if (!mbim_message_qdu_file_open_response_parse (
+							response,
+							&out_max_transfer_size,
+							&out_max_window_size,
+							&error)) {
+		g_debug ("error: couldn't parse response message: %s\n", error->message);
+		g_debug ("out_max_transfer_size: %d\n", (int)out_max_transfer_size);
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:361]: You should only attempt to use the contents of the out parameters if mbim_message_qdu_file_open_response_parse succeeds. If the method fails, you can only read the error, and you must ignore all the other fields.

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 32]
```diff
@@ -0,0 +1,545 @@
...
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	fu_device_set_progress_full (FU_DEVICE (ctx->device), (gsize) ctx->chunk_sent, (gsize) ctx->chunks->len);
+	ctx->chunk_sent++;
+	if (ctx->chunk_sent < ctx->chunks->len) {
+		FuChunk *chk = g_ptr_array_index (ctx->chunks, ctx->chunk_sent);
+		request = (mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), &error));
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:318]: Pass NULL at the GError, and assume set_new() never fails. Also, maybe drop the extra () around the method call

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 33]
```diff
@@ -0,0 +1,545 @@
...
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_write_response_parse (
+							 response,
+							 &error)) {
+		g_debug ("[%s][%d]: error: couldn't parse response message: %s\n", __FUNCTION__, __LINE__, error->message);
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:306]: Drop the __FUNCTION__ and __LINE__ from every log message in the patch please

File: plugins/modem-manager/fu-mbim-qdu-updater.c [hunk 34]
```diff
@@ -0,0 +1,545 @@
...
+			if (f) {
+				value = '1';
+				if (fwrite (&value, 1, 1, f) != 1) {
+					g_debug ("fail to write %s", rescan_path);
+				}
+			}
+
+			if (f)
+				fclose (f);
+			g_debug ("%s", remove_path);
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mbim-qdu-updater.c:250]: These logs don't seem useful at all, just printing paths without even saying what each path is. Looks to me more a debugging trace used during development than a proper debug log

File: plugins/modem-manager/fu-mbim-qdu-updater.c (hunk 15 above)
Comment (aleksander0m) [reply plugins/modem-manager/fu-mbim-qdu-updater.c:222]: Just think: the modem you're upgrading and is inhibited in MM and you can use it in fwupd, but there may be 30 additional MBIM modems also using the mbim-proxy, and you just broke the communication for all of them.

File: plugins/modem-manager/fu-plugin-modem-manager.c [hunk 35]
```diff
@@ -199,7 +215,8 @@ fu_plugin_mm_inhibit_device (FuPlugin *plugin, FuDevice *device, GError **error)
 
 	/* only do modem port monitoring using udev if the module is expected
 	 * to reset itself into a fully different layout, e.g. a fastboot device */
-	if (fu_mm_device_get_update_methods (FU_MM_DEVICE (device)) & MM_MODEM_FIRMWARE_UPDATE_METHOD_FASTBOOT) {
+	if (fu_mm_device_get_update_methods (FU_MM_DEVICE (device)) & MM_MODEM_FIRMWARE_UPDATE_METHOD_FASTBOOT ||
+	    fu_mm_device_get_update_methods (FU_MM_DEVICE (device)) & MM_MODEM_FIRMWARE_UPDATE_METHOD_MBIM_QDU) {
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-plugin-modem-manager.c:219]: I'm a bit confused. The module was in MBIM mode when managed as a modem, and the module is in MBIM mode during the upgrade process. Why do we need to handle the module as a udev device (which is an additional step required only when the modem "switches" between modem-mode and dowload-mode)?

All the work done in the patch to add support to the udev-device-based modem object doesn't seem to be needed at all.

File: plugins/modem-manager/fu-mm-device.c (hunk 10 above)
Comment (aleksander0m) [reply plugins/modem-manager/fu-mm-device.c:1027]: Also, following another comment in this review, the changes here are exclusively required when the layout of the module in the kernel changes (e.g. from modem-mode to download-mode, using a totally different layout). In your case, the module was MBIM in modem-mode and MBIM in download-mode, so there should be no need to handle it as a udev MM device.

File: plugins/modem-manager/fu-mm-device.c [hunk 36]
```diff
@@ -653,6 +728,86 @@ fu_mm_device_write_firmware_qmi_pdc (FuDevice *device, GBytes *fw, GArray **acti
...
+		return FALSE;
+	part = g_ptr_array_index (parts, 0);
+	filename = xb_node_get_attr (part, "filename");
+	csum = xb_node_get_attr (part, "MD5");
+	data = fu_archive_lookup_by_fn (archive, filename, error);
+	if (data == NULL)
+		return FALSE;
+	csum_actual = g_compute_checksum_for_bytes (G_CHECKSUM_MD5, data);
+	if (g_strcmp0 (csum, csum_actual) != 0) {
+		g_debug ("MD5 not matched");
```
Comment (aleksander0m) [inline plugins/modem-manager/fu-mm-device.c:799]: maybe include the name of the file that matched the MD5? that would improve the log a bit

File: contrib/fwupd.spec.in [hunk 37]
```diff
@@ -82,7 +82,7 @@ BuildRequires: flashrom-devel >= 1.2-2
 %endif
 
 %if 0%{?have_modem_manager}
-BuildRequires: ModemManager-glib-devel >= 1.10.0
+BuildRequires: ModemManager-glib-devel >= 1.17.0
```
Comment (aleksander0m) [inline contrib/fwupd.spec.in:85]: This needs to point to the MM dev version that includes the new MBIM_QDU symbol

**Review by aleksander0m — CHANGES_REQUESTED** [review]
It's a good start overall, but we definitely need to first work on solving the issues with the mbim-proxy first, and integrate the protocol bits in ModemManager as well.

**Comment by superm1** [comment]
Can we do something where we detect the MM version and only compile/support the new code if the new version is present?  
It takes time for distros to play catch up for all that stuff.

Otherwise we effectively are going to say that the new minimum requirement for MM plugin is this new unreleased version.

**Comment by aleksander0m** [comment]
> Can we do something where we detect the MM version and only compile/support the new code if the new version is present?
> It takes time for distros to play catch up for all that stuff.
> 
> Otherwise we effectively are going to say that the new minimum requirement for MM plugin is this new unreleased version.

Oh, that is a very good idea, yes. For this new code, once the new MBIM_QDU symbol is added to git master I'll update the dev version to e.g. 1.17.1 and then we can do `#if MM_CHECK_VERSION (1,17,1) ...... #endif`

File: plugins/modem-manager/fu-mm-device.c (hunk 8 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mm-device.c:126]: > Could you please push the MM changes first, we get them merged in MM, we then bump the MM dev version including the changes (e.g. to 1.17.1 or whatever) and then we build-depend on that version for this integration in fwupd?

Sure, the qdu related parts in MM will be pushed soon, if no problems, please merged it in MM, thanks.

File: plugins/modem-manager/fu-mm-device.c (hunk 9 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mm-device.c:318]: > but it would probably be worth getting merged to make it work also with this new QDU method: [aleksander0m@abf7685](https://github.com/aleksander0m/fwupd/commit/abf7685e8e233be01da4f677508f88a690ff7ecf)
> 
> Could you see if that patch helps you to do the same?

It looks great in your patch, could you tell me when it can be merged in?

File: plugins/modem-manager/fu-mbim-qdu-updater.h (hunk 7 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mbim-qdu-updater.h:17]: Yes, I will change it, thanks a lot.

File: plugins/modem-manager/fu-mbim-qdu-updater.c (hunk 15 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mbim-qdu-updater.c:222]: Yes, you are right, the mbim-proxy problem need to be fixed rather than be killed, I will not kill it here.

The remove and rescan operations used because when module reboot after download firmware, the driver will not initialize successfully, maybe it is our hardware defects. I will get rid of remove and rescan here, what do you think? 
Or could you give me some suggestion?  Thank you very much.

File: plugins/modem-manager/fu-mbim-qdu-updater.c (hunk 26 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mbim-qdu-updater.c:421]: Ok, thanks for your suggestion.

File: plugins/modem-manager/fu-mbim-qdu-updater.c (hunk 15 above)
Comment (aleksander0m) [reply plugins/modem-manager/fu-mbim-qdu-updater.c:222]: I would remove this logic from the patch, and try to see how to solve it, maybe it's a kernel driver issue that is better fixed in the kernel instead of via workarounds in userspace?

File: plugins/modem-manager/fu-mm-device.c (hunk 9 above)
Comment (aleksander0m) [reply plugins/modem-manager/fu-mm-device.c:318]: I know there is at least one additional team working on top of that patch, so maybe it's worth merging that already so that everyone can work on top of it? @hughsie what do you think? should I suggest that single patch (adding generic PCI support for the MM plugin) in a separate pull request?

File: plugins/modem-manager/fu-mbim-qdu-updater.c (hunk 15 above)
Comment (aleksander0m) [reply plugins/modem-manager/fu-mbim-qdu-updater.c:222]: > Yes, you are right, the mbim-proxy problem need to be fixed rather than be killed, I will not kill it here.

If there is any problem with the mbim-proxy, please report it as an issue to libmbim including verbose logs of the proxy itself, see https://modemmanager.org/docs/modemmanager/debugging/#qmi-proxy-or-mbim-proxy-debug-logs. I'd love to help fixing that problem!

File: plugins/modem-manager/fu-mm-device.c (hunk 8 above)
Comment (aleksander0m) [reply plugins/modem-manager/fu-mm-device.c:126]: I have now merged to git master in MM, and pushed a new version bump to 1.17.1, so, can you rework your changes so that the MBIM QDU support is included in between `#if MM_CHECK_VERSION (1,17,1) ...   #endif` checks?

File: plugins/modem-manager/fu-mm-device.c (hunk 9 above)
Comment (aleksander0m) [reply plugins/modem-manager/fu-mm-device.c:318]: I have created this pull request with that commit after rebasing them on top of git master: https://github.com/fwupd/fwupd/pull/3210

File: plugins/modem-manager/fu-mbim-qdu-updater.c (hunk 15 above)
Comment (Jarvis-Jiang-G) [reply plugins/modem-manager/fu-mbim-qdu-updater.c:222]: > If there is any problem with the mbim-proxy, please report it as an issue to libmbim including verbose logs of the proxy itself,

I have report the issue to libmbim including verbose logs of the mbim-proxy, https://gitlab.freedesktop.org/mobile-broadband/libmbim/-/issues/17. Thank you for helping me fix this problem.

**Comment by Jarvis-Jiang-G** [comment]
Thanks for your revision suggestions @aleksander0m @hughsie. And I have modified code according to your suggestion, changes including
1. Some code format problems
2. Based on the new patch (#3210 ) to make a modification of code logic
3. Remove some special operations, including remove and rescan device, kill mbim-proxy
4. Module is in MBIM mode during the upgrade process, so no longer handle it as a udev device

I have recreated a new MR in #3222 , thanks.

**Comment by superm1** [comment]
Close in favor of #3222
